8293336: AOT-linking of invokedynamic for lambda expression and string concat

iklam opened the pull request:
This is a draft pr. Details TBD.

liach commented on src/java.base/share/classes/java/lang/invoke/MethodType.java:
```diff
@@ -398,6 +400,13 @@ private static MethodType makeImpl(Class<?> rtype, Class<?>[] ptypes, boolean tr
         }
         MethodType primordialMT = new MethodType(rtype, ptypes);
+        if (AOTHolder.archivedMethodTypes != null) {
+            MethodType mt = AOTHolder.archivedMethodTypes.get(primordialMT);
```
Can we populate these AOT MTs into the internTable so we only look up once?

iklam replied:
There's a trade off here -- start-up will be slowed down if the archivedMethodTypes table is big. Also, `internTable` uses more memory as it wraps every entry inside a weak reference.

turbanoff commented on test/lib/jdk/test/lib/cds/CDSTestUtils.java:
```diff
@@ -863,15 +863,20 @@ public static Path copyFile(String srcFile, String destDir) throws Exception {
+            }
         }
-        if (javaopts != null && javaopts.contains("-XX:+AOTClassLinking")) {
+        if (lastMatch != null && lastMatch.equals("-XX:+" +  optionName)) {
```
```suggestion
        if (lastMatch != null && lastMatch.equals("-XX:+" + optionName)) {
```

shipilev commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -318,6 +325,201 @@ void AOTConstantPoolResolver::maybe_resolve_fmi_ref(InstanceKlass* ik, Method* m
+
+  // We currently support only StringConcatFactory::makeConcatWithConstants() and LambdaMetafactory::metafactory()
+
+  if (bsm_klass->equals("java/lang/invoke/StringConcatFactory") &&
+      bsm_name->equals("makeConcatWithConstants")) {
+    return true;
```
I think all BSM entries in SCF are deterministic, really. So we can just check for the class?

iklam replied:
We haven't done much testing with the other SCF BSMs, so I think it's better to do that in a follow-up REF.

ExE-Boss replied:
`StringConcatFactory​::makeConcat(…)` is definitely deterministic, as it simply delegates to `StringConcatFactory​::makeConcatWithConstants(…)` with the recipe being based on the number of `MethodType` parameters:
https://github.com/openjdk/jdk/blob/de90204b60c408ef258a2d2515ad252de4b23536/src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java#L231-L240

And there is no other bootstrap method in `StringConcatFactory` at the moment.

iklam replied:
Since the comment says "This bootstrap method is unlikely to be used in practice", I am not sure if we can sufficiently test it (with aot-linking).

shipilev commented on src/hotspot/share/cds/aotClassInitializer.cpp:
```diff
@@ -34,11 +35,18 @@ bool AOTClassInitializer::can_archive_initialized_mirror(InstanceKlass* ik) {
+  }
+
+  Symbol* name = ik->name();
+  if (name->equals("jdk/internal/constant/PrimitiveClassDescImpl") ||
```
Could we / should we start using `vmSymbols` constants for these? I think you can directly compare the `Symbol*`-s then, like `name == vmSymbols::jdk_internal_constant_PrimitiveClassDescImpl_Holder() `.

iklam replied:
Not every class in this table is inside vmSymbols, and I don't want to have to edit vmSymbols whenever a new class is added here (and having to do garbage collection in vmSymbols when a class is removed from here).

I pushed a new version that's more table driven, and avoid most strcmp calls by filtering with the symbol length.

shipilev reviewed (commented):
Cursory review:

adinn commented on src/hotspot/share/cds/aotConstantPoolResolver.cpp:
```diff
@@ -318,6 +327,201 @@ void AOTConstantPoolResolver::maybe_resolve_fmi_ref(InstanceKlass* ik, Method* m
+  }
+
+  Symbol* mh_sig = cp->method_handle_signature_ref_at(mh_index);
+  if (!check_lambda_metafactory_signature(cp, mh_sig, false)) {
```
This appears to be rerunning the previous check without a log message -- or am I missing something here?

iklam replied:
It was a cut-and-paste mistake. I've removed this line in [195303f](https://github.com/openjdk/jdk/pull/21143/commits/195303f2c6939b93b6152c6610bd38faff5ae3c7) , where I also fixed some minor bugs with `AOTConstantPoolResolver::is_indy_resolution_deterministic()` and added more comments in the C++ code about what checks are being made.

I also added tests code in AOTLinkedLambdas.java to cover all the type checks that happen inside `is_indy_resolution_deterministic()`.

adinn commented on src/hotspot/share/cds/archiveBuilder.cpp:
```diff
@@ -228,6 +230,9 @@ bool ArchiveBuilder::gather_klass_and_symbol(MetaspaceClosure::Ref* ref, bool re
     if (!is_excluded(klass)) {
       _klasses->append(klass);
+      if (klass->is_hidden() && klass->is_instance_klass()) {
+        update_hidden_class_loader_type(InstanceKlass::cast(klass));
```
Can you explain why this 'update' is needed. Are the shared classloader type and classpath index not already set? Or do they need adjusting? Perhaps a comment would help.

iklam replied:
Thanks for pointing this out. This code was necessary because `ClassLoader::record_result()` was not handling hidden classes. I fixed this by moving the logic from `update_hidden_class_loader_type()` to `ClassLoader::record_hidden_class()`.

adinn commented on src/hotspot/share/cds/cdsConfig.cpp:
```diff
@@ -473,6 +481,11 @@ bool CDSConfig::check_vm_args_consistency(bool patch_mod_javabase, bool mode_fla
 
+bool CDSConfig::allow_only_single_java_thread() {
+  // See comments in JVM_StartThread()
+  return is_dumping_static_archive();
```
The test in `JVM_StartThread()` calls `CDSConfig::is_dumping_static_archive()`. Should it be updated to call `CDSConfig::allow_only_single_java_thread()`?

iklam replied:
I updated `JVM_StartThread()`  to call `CDSConfig::allow_only_single_java_thread()`

adinn commented on src/hotspot/share/cds/cdsHeapVerifier.cpp:
```diff
@@ -124,6 +126,13 @@ CDSHeapVerifier::CDSHeapVerifier() : _archived_objs(0), _problems(0)
   ADD_EXCL("sun/invoke/util/ValueConversions",           "ONE_INT",                // E
```
At line 46 the example that explains how the verifier works discusses save and restore of field `Foo#archivedFoo`. I believe that field was supposed to be declared as `static` but it is actually declared as an instance field.

iklam replied:
I edited the comment to add `static` to the declaration of `Foo#archivedFoo`.

adinn commented on src/hotspot/share/cds/cdsHeapVerifier.cpp:
```diff
@@ -232,10 +265,19 @@ inline bool CDSHeapVerifier::do_entry(oop& orig_obj, HeapShared::CachedOopInfo&
     ResourceMark rm;
+    char* class_name = info->_holder->name()->as_C_string();
+    char* field_name = info->_name->as_C_string();
+    if (strstr(class_name, "java/lang/invoke/BoundMethodHandle$Species_") == class_name &&
```
Can we have a comment to explain this special case?

iklam replied:
This check is no longer necessary, the Species_XXX classes are now added to the aot-init list in `AOTClassInitializer::can_archive_initialized_mirror()`. I've removed these two lines.

adinn commented on src/hotspot/share/cds/classListParser.cpp:
```diff
@@ -597,6 +597,18 @@ void ClassListParser::resolve_indy(JavaThread* current, Symbol* class_name_symbo
+  // This is an older CDS optimization:
+  // We store a pre-generated version of the lambda proxy class in the AOT cache,
+  // which will be loaded via JVM_LookupLambdaProxyClassFromArchive().
+  // This eliminate dynamic class generation of the proxy class, but we still need to
```
"This eliminate" --> "This eliminates"

iklam replied:
Fixed

adinn commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -755,6 +889,10 @@ void KlassSubGraphInfo::add_subgraph_object_klass(Klass* orig_k) {
 
 void KlassSubGraphInfo::check_allowed_klass(InstanceKlass* ik) {
+  if (CDSConfig::is_dumping_invokedynamic()) {
+    // We allow LambdaProxy classes in platform and app loaders as well.
```
Do we not want an assert here to cover the extra classes we are allowing?

iklam replied:
I updated `KlassSubGraphInfo::check_allowed_klass()` to do proper checks for lambda proxy classes that are allowed. I also update the comment at at line 1975 to be consistent with `check_allowed_klass()`.

ExE-Boss reviewed (commented):
<!---->

DanHeidinga commented on src/java.base/share/classes/java/util/concurrent/ConcurrentHashMap.java:
```diff
@@ -595,7 +596,15 @@ public class ConcurrentHashMap<K,V> extends AbstractMap<K,V>
 
     /** Number of CPUS, to place bounds on some sizings */
-    static final int NCPU = Runtime.getRuntime().availableProcessors();
+    static @Stable int NCPU;
```
I would prefer to not mark this `@Stable` at this time as it would have different assembly and runtime values and instead add a followup RFE to investigate adding it in later.

iklam replied:
We have been archiving `ConcurrentHashMap` objects for many JDK releases (to support archiving of modules) and `NCPU` would change between CDS archive dump time and program run time. We have not seen any problems, so I think it's safe to use the `@Stable` annotation.

stefank commented on src/hotspot/share/memory/iterator.inline.hpp:
```diff
@@ -46,7 +46,11 @@ inline MetadataVisitingOopIterateClosure::MetadataVisitingOopIterateClosure(Refe
 
 inline void ClaimMetadataVisitingOopIterateClosure::do_cld(ClassLoaderData* cld) {
-  cld->oops_do(this, _claim);
+  if (cld != nullptr) {
+    // Could be null during early VM bootstrap for archived heap objects whose
+    // class has not yet been loaded by CDS.
+    cld->oops_do(this, _claim);
+  }
```
Could you show the stack trace for this? I want to understand better why we are iterating over objects that don't have a loaded class.

iklam replied:
Here's the call stack if I remove the "if" check on line 49. It can be reproduced with test/hotspot/jtreg/runtime/cds/appcds/aotClassLinking/BulkLoaderTest.java from the PR repo.

```
10 javaSignalHandler (sig=11, info=0x7ffff428fdb0, context=0x7ffff428fc80)
12 Atomic::PlatformLoad::operator() (this=0x7ffff429097f, dest=0x28)
13 Atomic::LoadImpl, void>::operator() (this=0x7ffff42909af, dest=0x28)
14 Atomic::load (dest=0x28)
15 ClassLoaderData::try_claim (this=0x0, claim=4)
16 ClassLoaderData::oops_do (this=0x0, f=0x7fff6c014c88, claim_value=4, clear_mod_oops=false)
17 ClaimMetadataVisitingOopIterateClosure::do_cld (this=0x7fff6c014c88, cld=0x0)
18 ClaimMetadataVisitingOopIterateClosure::do_klass (this=0x7fff6c014c88, k=0x7fff763655d8)
19 call_do_klass (closure=0x7fff6c014c88, k=0x7fff763655d8)
20 Devirtualizer::do_klass (closure=0x7fff6c014c88, k=0x7fff763655d8)
21 InstanceKlass::oop_oop_iterate (closure=0x7fff6c014c88, obj=0xfffd72a0, this=0x7fff763655d8)
22 OopOopIterateDispatch::Table::oop_oop_iterate (cl=0x7fff6c014c88, obj=0xfffd72a0, k=0x7fff763655d8)
23 OopIteratorClosureDispatch::oop_oop_iterate (cl=0x7fff6c014c88, obj=0xfffd72a0, klass=0x7fff763655d8)
24 oopDesc::oop_iterate (this=0xfffd72a0, cl=0x7fff6c014c88)
25 G1FullGCMarker::follow_object (this=0x7fff6c014630, obj=0xfffd72a0)
26 G1FullGCMarker::publish_and_drain_oop_tasks (this=0x7fff6c014630)
27 G1FullGCMarker::follow_marking_stacks (this=0x7fff6c014630)
28 G1FullGCMarker::complete_marking (this=0x7fff6c014630, oop_stacks=0x7fffd445c6b8, array_stacks=0x7fffd445c6d0, terminator=0x7fffd445c190)
29 G1FullGCMarkTask::work (this=0x7fffd445c030, worker_id=0)
30 WorkerTaskDispatcher::worker_run_task (this=0x7ffff008b9b8)
31 WorkerThread::run (this=0x7ffff008bac0)
```

This is in a GC triggered by here (with -XX:VerifyArchivedFields=2).

```
void HeapShared::init_archived_fields_for(Klass* k, 
     const ArchivedKlassSubGraphInfoRecord* record) {
  verify_the_heap(k, "before");
```

At frame 18:

```
#18 0x00007ffff58505d1 in ClaimMetadataVisitingOopIterateClosure::do_klass 
    (this=0x7fff6c014c88, k=0x7fff763655d8)
    at src/hotspot/share/memory/iterator.inline.hpp:58
58	  ClaimMetadataVisitingOopIterateClosure::do_cld(cld);
(gdb) psym k->_name
$1 = 0x7fff765430be "BulkLoaderTestApp$$Lambda+0x800000013"
(gdb) p k->_class_loader_data
$2 = (ClassLoaderData *) 0x0
(gdb) p SystemDictionary::_java_system_loader 
$3 = {_obj = 0x0}
```

At frame 21:

```
#21 0x00007ffff6132f6f in InstanceKlass::oop_oop_iterate<narrowOop, G1MarkAndPushClosure> 
    (closure=0x7fff6c014c88, obj=0xfffd72a0, this=0x7fff763655d8)
    at src/hotspot/share/oops/instanceKlass.inline.hpp:164
164	    Devirtualizer::do_klass(closure, this);
(gdb) call pp(obj)
"Executing pp"
BulkLoaderTestApp$$Lambda/0x800000013 
{0x00000000fffd72a0} - klass: 'BulkLoaderTestApp$$Lambda+0x800000013' - flags: is_hidden_class 
```

The root cause is:

- A lambda has been aot-resolved for a class loaded by the app loader
- There is an instance (`0xfffd72a0`) of the lambda proxy class (`BulkLoaderTestApp$$Lambda+0x800000013`) that represents this call site 
- At the point of failure, all archived heap objects, including `0xfffd72a0`, are reachable (through `HeapShared::_root_segments`).
- However, we are still early in VM bootstrap. The `_java_system_loader` isn't available yet, so we are not yet able to load the lambda proxy class.

Before JEP 483, all archived objects are from java.base, and the `VerifyArchivedFields` code was careful to not trigger a GC until `is_init_completed()` becomes true.

However, with JEP 483, it's possible to have an archived object from the app class loader, and it's legally possible for a GC to happen before the app loader is available. Thus we come into this situation.

(Ping @fisk @coleenp in case you're curious)

iklam replied:
I filed [JDK-8342429](https://bugs.openjdk.org/browse/JDK-8342429) "Cache classes in loaded state" to address this properly.

stefank replied:
Thanks!

stefank commented on src/hotspot/share/oops/cpCache.cpp:
```diff
@@ -49,6 +49,7 @@
 #include "oops/method.inline.hpp"
 #include "oops/objArrayOop.inline.hpp"
 #include "oops/oop.inline.hpp"
+#include "oops/method.inline.hpp"
```
Could you go over the entire patch and fix sort order of added includes.

iklam replied:
I fixed the include problems found by my script. I'll make sure to run the script again before the final push.

liach commented:
I recommend adding javadoc or block comments indicating that `runtimeSetup`, `resetArchivedStates`, etc. are called by VM CDS code so people don't accidentally rename this and have unintended impact.

liach commented on src/java.base/share/classes/java/lang/invoke/MethodType.java:
```diff
@@ -398,6 +400,13 @@ private static MethodType makeImpl(Class<?> rtype, Class<?>[] ptypes, boolean tr
         }
         MethodType primordialMT = new MethodType(rtype, ptypes);
+        if (archivedMethodTypes != null) {
+            MethodType mt = archivedMethodTypes.get(primordialMT);
```
The use of `HashMap` naturally feels alarming given this method is invoked in a concurrent context. I just want to make sure that:

1. `HashMap` will never have structural modifications when read access is performed (some data structures like treap do). Don't know if we have a test to verify such an intended usage, as we would hurt badly if this implied property suddenly no longer exist.
2. The `archivedMethodTypes` must be properly published so we won't read a partial `HashMap`. I've heard of talks about concurrent VM bootstraping so I need some help to understand how this is ensured in bootstrap.

iklam replied:
I added asserts to check that the HashMap is written only in the CDS dumping process, so at least we know it's not being modified when we read from it.

The CDS archived heap is either mmaped or read into memory while the VM is still in single threaded mode, so the entire content of the archived heap should be published for to all future threads.

iklam commented on src/hotspot/share/cds/aotLinkedClassBulkLoader.cpp:
```diff
@@ -233,6 +232,31 @@ void AOTLinkedClassBulkLoader::initiate_loading(JavaThread* current, const char*
   }
 }
 
+// TODO -- is this really correct? Do we need a special ClassLoaderData for each hidden class?
```
We only resolve indys for the boot/platform/app class loaders, so there's no need to support class unloading. I've added a comment and tightened up the code with asserts. See version [382446d](https://github.com/openjdk/jdk/pull/21143/commits/382446d41bb79c4bd2c1dfeda656dfa2a73466b4)

iklam replied:
(I've replied to this comment. I can see it in the mailing list but it seems to be lost on this page, so I'll add it again)

We only resolve indys for the boot/platform/app class loaders, so there's no need to support class unloading. I've updated the comment and tghtened up the code with asserts. See version [382446d](https://github.com/openjdk/jdk/pull/21143/commits/382446d41bb79c4bd2c1dfeda656dfa2a73466b4)

liach replied:
This is correct for the LMF, but I think we are using weak hidden classes in string concatenation.

DanHeidinga replied:
I just confirmed - the `InlineHiddenClassStrategy` in j.l.i.StringConcatFactory uses weak hidden classes.  It looks like there are two possible strategies though, and the InlineHiddenClassStrategy is only used if the MH strategy fails (more parameters than the high arity threshold)

liach replied:
Currently the high arity threshold is 0, meaning weak hidden classes are always used.

liach commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -443,8 +443,13 @@ bool HeapShared::is_lambda_proxy_klass(InstanceKlass* ik) {
 }
 
+bool HeapShared::is_string_concat_klass(InstanceKlass* ik) {
+  return ik->is_hidden() && ik->name()->starts_with("java/lang/String$$StringConcat");
```
Can you add a comment in `StringConcatFactory`'s place where this name is defined, so others who change the generated class name will remember to update references here?

iklam replied:
Done.

ashu-mehra commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -601,6 +641,104 @@ void HeapShared::set_has_native_pointers(oop src_obj) {
+  }
+};
+
+void HeapShared::find_archivable_hidden_classes_in_object(oop root) {
```
If this function find archivable hidden classes, shouldn't there be a check for`is_hidden()` on the InstanceKlass being marked as `required`. Am I missing something?

iklam replied:
I added the `ik->is_hidden()` check as you suggested. I also renamed the `find_archivable_hidden_classes_xxx` functions to `find_required_hidden_classes_xxx()`. This is to avoid confusion with `HeapShared::is_archivable_hidden_klass()`, which means "it is possible to archive X", but not "it is necessary to archive X".

ashu-mehra commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -647,16 +785,32 @@ void HeapShared::copy_interned_strings() {
-  archive_java_mirrors();
-  archive_strings();
-  Universe::archive_exception_instances();
+  {
```
Why was this scope introduced?

iklam replied:
To indicate that these calls must be made while the "see objects" table is available.

ashu-mehra commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -629,7 +645,83 @@ class UnregisteredClassesDuplicationChecker : StackObj {
+}
+
+// Returns true if the class should be excluded. This can be called before
+// SystemDictionaryShared::check_excluded_classes().
```
There are a couple of references to check_excluded_classes() in the comments but this function no longer exists. Can you please update the comment accordingly.

iklam replied:
I replaced with the new name of the function `find_all_archivable_classes()`.

ashu-mehra commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -642,10 +734,50 @@ void SystemDictionaryShared::check_excluded_classes() {
+    auto check_hidden = [&] (InstanceKlass* k, DumpTimeClassInfo& info) {
+      if (k->is_hidden() && should_hidden_class_be_archived(k)) {
+        SystemDictionaryShared::check_for_exclusion(k, &info);
+        if (info.is_excluded()) {
```
Is it possible that a hidden class for which `should_hidden_class_be_archived()` returns true is marked for exclusion by `SDS::check_for exclusion`? I guess only possibility is if for some reason its super or interface is excluded. Is that possible?

iklam replied:
We filter out hidden classes that cannot be supported (e.g., when a Lambda uses an interface using "old" bytecodes). See

https://github.com/openjdk/jdk/pull/21143/files#diff-fc60af483d061250f01054372fd56eba517a6b8dd37a742a66dfb0e15cb68e40R366-R372

https://github.com/openjdk/jdk/pull/21143/files#diff-2520dedd703ec600cd3037ec0008e9d7f45c6a41acaefac019339641bc749163R149-R152

ashu-mehra commented on src/hotspot/share/classfile/systemDictionaryShared.cpp:
```diff
@@ -642,10 +734,50 @@ void SystemDictionaryShared::check_excluded_classes() {
+
+  // Now, all hidden classes that have not yet been scanned must be marked as excluded
+  auto exclude_remaining_hidden = [&] (InstanceKlass* k, DumpTimeClassInfo& info) {
+    if (k->is_hidden() && !info.has_checked_exclusion()) {
```
Suggestion: Check for `has_checked_exclusion()` is not really required as the call to `SDS::check_for_exclusion()` does that. And the condition in `guarantee()` can be updated as `info.is_excluded() || info.is_required()`.

iklam replied:
I changed the code to the following to make sure that `should_hidden_class_be_archived(k)` is mutually exclusive with `info.is_excluded()`.

```
  auto exclude_remaining_hidden = [&] (InstanceKlass* k, DumpTimeClassInfo& info) {
    if (k->is_hidden()) {
      SystemDictionaryShared::check_for_exclusion(k, &info);
      if (CDSConfig::is_dumping_invokedynamic()) {
        if (should_hidden_class_be_archived(k)) {
          guarantee(!info.is_excluded(), "Must be");
        } else {
          guarantee(info.is_excluded(), "Must be");
        }
      }
    }
  };
```

Unfortunately I can do this only for `CDSConfig::is_dumping_invokedynamic()`. In the "legacy CDS" mode of lambda support, we actually would return true from `should_hidden_class_be_archived(k)` and then later decide to exclude `k`.

ashu-mehra commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -601,6 +641,104 @@ void HeapShared::set_has_native_pointers(oop src_obj) {
+    ArchivableStaticFieldInfo* info = &fields[i];
+    const char* klass_name = info->klass_name;
+    for (; fields[i].valid(); i++) {
+      ArchivableStaticFieldInfo* f = &fields[i];
```
It looks like the pattern of nested for-loops is copied from HeapShared::archive_object_subgraphs()
but this pattern doesn't seem to give any advantage here. If so can we replace it with just single for loop to make it easier to understand?

ashu-mehra commented on src/hotspot/share/oops/constantPool.cpp:
```diff
@@ -300,8 +330,15 @@ objArrayOop ConstantPool::prepare_resolved_references_for_archiving() {
     ConstantPool* src_cp = ArchiveBuilder::current()->get_source_addr(this);
+    src_cp->iterate_archivable_resolved_references([&](int rr_index) {
+      keep_resolved_refs.at_put(rr_index, true);
+      });
```
Indentation seems off here

iklam replied:
Fixed.

ashu-mehra commented:
@iklam Changes to move main module name from modules.cpp to elsewhere do not look related to AOT-linking of indy. If so, can they be done in a separate PR.

ashu-mehra commented on src/hotspot/share/cds/dumpTimeClassInfo.hpp:
```diff
@@ -42,7 +42,8 @@ class DumpTimeClassInfo: public CHeapObj<mtClass> {
   bool                         _is_early_klass;
   bool                         _has_checked_exclusion;
-
+  bool                         _is_required;
```
Is this flag only used for hidden classes?

iklam replied:
I changed it to `_is_required_hidden_class`.

iklam commented:
> I recommend adding javadoc or block comments indicating that `runtimeSetup`, `resetArchivedStates`, etc. are called by VM CDS code so people don't accidentally rename this and have unintended impact.

Done.

rose00 commented:
I read through it fairly thoroughly.  I made a lot of suggestions.  Rather than edit comments into lots of spots in this PR, I made the following patch file.
[aot-indy-21143-commentary.diff.txt](https://github.com/user-attachments/files/17408940/aot-indy-21143-commentary.diff.txt)

  The first lines of the patch file state which revision the patch applies to, which is 4be6a255cdd from Tuesday.

dholmes-ora commented:
> I read through it fairly thoroughly. I made a lot of suggestions. Rather than edit comments into lots of spots in this PR, I made the following patch file. [aot-indy-21143-commentary.diff.txt](https://github.com/user-attachments/files/17408940/aot-indy-21143-commentary.diff.txt)

```
+bool InstanceKlass::is_enum_subclass(bool direct_only) const {
+  InstanceKlass* s = java_super();
+  return (s == vmClasses::Enum_klass() ||
+          (s != null && s->java_super() == vmClasses::Enum_klass()));
+}
```
Just happened to notice you aren't checking `direct_only`.

iklam commented on src/hotspot/share/cds/metaspaceShared.cpp:
```diff
@@ -841,6 +853,26 @@ void MetaspaceShared::preload_and_dump_impl(StaticArchiveBuilder& builder, TRAPS
+                             vmSymbols::void_method_signature(),
+                             CHECK);
+
+      // java.lang.Class::reflectionFactory cannot be archived yet. We set this field
+      // to null, and it will be initialized again at runtime.
+      log_debug(cds)("Resetting Class::reflectionFactory");
+      TempNewSymbol method_name = SymbolTable::new_symbol("resetArchivedStates");
+      Symbol* method_sig = vmSymbols::void_method_signature();
+      JavaCalls::call_static(&result, vmClasses::Class_klass(),
+                             method_name, method_sig, CHECK);
```
`Class::reflectionFactory` is a large table that must be cleared, or else we will bring unwanted (unsupported) objects into the cache.

During runtime, `reflectionFactory` will be initialized on demand, so there's no need to explicitly create it with `runtimeSetup`.

liach replied:
That said, can reflectionFactory field be stable in the future with this?

iklam replied:
Perhaps. In the Leyden repo we are doing some optimizations for reflection. See the `ArchiveReflectionData` flag in the Leyden premain repo.. I am not sure if `reflectionFactory` itself needs to be cached.

iklam commented:
> I read through it fairly thoroughly. I made a lot of suggestions. Rather than edit comments into lots of spots in this PR, I made the following patch file. [aot-indy-21143-commentary.diff.txt](https://github.com/user-attachments/files/17408940/aot-indy-21143-commentary.diff.txt)
> 
> The first lines of the patch file state which revision the patch applies to, which is [4be6a25](https://github.com/openjdk/jdk/commit/4be6a255cdd300bb6e15ded1dc163b00cd24dbd8) from Tuesday.

Hi John,

Thank you so much for the code. I have merged most of your diffs into the PR. There are a few items left which I will go over tomorrow. I simplified the `InstanceKlass::interface_needs_clinit_execution_as_super()` as I don't need to check against non-interface classes, so I worry about adding code that's not used by this JEP.

I also moved the `InstanceKlass::assert_no_clinit_will_run_for_aot_initialized_class()` function to be closed to `InstanceKlass::initialize_with_aot_initialized_mirror()` -- the latter has been simplified as we would never need to take the "quicker path" from the previous version of the code.

I also cleaned up the "firewall" check in `mark_for_aot_initialization()`.

iklam commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -806,6 +1122,25 @@ bool KlassSubGraphInfo::is_non_early_klass(Klass* k) {
+    if (!is_trivial_clinit(interfaces->at(i))) {
+      return false;
+    }
+  }
```
I've removed this function altogether. It was used in logging only, and it wasn't clear what the output meant (w.r.t to interfaces that have non-trivial <clinit>). I think it's best to remove this function, and implement it properly in a future RFE that automatically aot-init all classes that can be initialized with no observable side effects.

stefank commented on src/hotspot/share/memory/iterator.hpp:
```diff
@@ -172,6 +173,8 @@ class ClaimingCLDToOopClosure : public CLDToOopClosure {
 };
 
 class ClaimMetadataVisitingOopIterateClosure : public OopIterateClosure {
+ private:
+  static void assert_is_pending_aot_linked_class(Klass* k) NOT_DEBUG_RETURN;
```
```suggestion
 private:
  static void assert_is_pending_aot_linked_class(Klass* k) NOT_DEBUG_RETURN;

```

stefank commented on src/hotspot/share/memory/iterator.inline.hpp:
```diff
@@ -51,7 +51,11 @@ inline void ClaimMetadataVisitingOopIterateClosure::do_cld(ClassLoaderData* cld)
+  if (cld != nullptr) {
+    ClaimMetadataVisitingOopIterateClosure::do_cld(cld);
+  } else {
+    assert_is_pending_aot_linked_class(k);
```
Would there be a drawback to just adding `assert(AOTLinkedClassBulkLoader::is_pending_aot_linked_class(k), "sanity");` here and skip the changes the both .hpp and .cpp files?

iklam replied:
I changed the code to `assert(is_pending_aot_linked_class(k), "sanity")`. Since iterator.hpp is widely used, I avoid including aotLinkedClassBulkLoader.hpp in this header, and route the call through iterator.cpp.

stefank replied:
My proposal was to use it in this file (iterator.inline.hpp) and not iterator.hpp. I see no need to route this through the .cpp file.

iklam replied:
iterator.inline.hpp is included by many cpp file (580 out of 1409 .o files in hotspot), so I don't want to add aotLinkedClassBulkLoader.hpp to it.

```
scanning    580 iterator.inline.hpp
   2 found    580 oop.inline.hpp
   3 found    580 zIterator.inline.hpp
   4 found    580 zBarrierSet.inline.hpp
   5 found     60 g1OopClosures.inline.hpp
   6 found      7 g1FullGCOopClosures.inline.hpp
   7 found      5 psPromotionManager.inline.hpp
```

stefank replied:
From what I can tell it is a tiny header with very few dependencies. I don't think we gain much by trying to prevent it from being include. I would strongly prefer to have cleaner code in iterator.hpp/cpp than trying to get rid of that include.

iklam replied:
I have moved the aotLinkedClassBulkLoader.hpp to iterator.inline.hpp as you suggested:

https://github.com/openjdk/jdk/pull/21642/files#diff-9512406aa5bffbef807aa26007f54e2bfb6b0635e1829aa77bf483bdfd342c8cR30

iklam commented:
> ```
> +bool InstanceKlass::is_enum_subclass(bool direct_only) const {
> +  InstanceKlass* s = java_super();
> +  return (s == vmClasses::Enum_klass() ||
> +          (s != null && s->java_super() == vmClasses::Enum_klass()));
> +}
> ```
> 
> Just happened to notice you aren't checking `direct_only`.

I consulted with John and removed the `direct_only` parameter as it's no longer needed.

iklam commented:
> @iklam Changes to move main module name from modules.cpp to elsewhere do not look related to AOT-linking of indy. If so, can they be done in a separate PR.

The change is actually necessary --  the main module name used to be restored inside `MetaspaceShared::serialize()`, which happens after we decide to accept the CDS archive. When the main module doesn't match, we will disable the "archived full module graph" and keep on using the rest of the CDS archive.

With aot-linked classes, the main module is needed early to decide whether to reject a CDS archive altogether. Therefore, I needed to lift its processing our of `MetaspaceShared::serialize()` and put its information directly inside the archive file header.

(This change could have been included in https://github.com/openjdk/jdk/pull/20843 , but that might be too much work now that the review for JEP 483 is winding down).

iklam commented on src/hotspot/share/cds/archiveUtils.cpp:
```diff
@@ -371,6 +371,13 @@ void ArchiveUtils::log_to_classlist(BootstrapInfo* bootstrap_specifier, TRAPS) {
+bool ArchiveUtils::has_aot_initialized_mirror(InstanceKlass* src_ik) {
+  if (SystemDictionaryShared::is_excluded_class(src_ik)) {
+    return false;
+  }
```
These classes are not buffered so `ArchiveBuilder::current()->get_buffered_addr(src_ik)` won't work. I added this assert instead.

```
  if (SystemDictionaryShared::is_excluded_class(src_ik)) {
    assert(!ArchiveBuilder::current()->has_been_buffered(src_ik), "sanity");
    return false;
  }
```

iklam commented on src/hotspot/share/cds/heapShared.cpp:
```diff
@@ -431,12 +432,33 @@ void HeapShared::remove_scratch_objects(Klass* k) {
+
+  if (HeapShared::is_archivable_hidden_klass(ik)) {
+    // We can't rerun the <clinit> method of hidden classes as we don't save
+    // the classData, so we must archive its mirror in initialized state.
```
This comment is obsolete. Also, we won't come for any class unless it's already initialized. So I deleted the comment and changed the assert to this for all classes

```
assert(ik->is_initialized(), "must be");
```

iklam commented:
Closing this PR as its changes have been combined into https://github.com/openjdk/jdk/pull/21642.
